### PR TITLE
[Bugfix][V1] Gate async-scheduling output-token-id repair on a declared logitsproc capability

### LINKED ISSUE #52461: [Bug]: Under async scheduling, stateful logits processors that read output_token_ids values see -1 placeholders — logitsprocs_need_output_token_ids only counts CLI custom processors
### Your current environment

Observed on `releases/v0.27.1` (`6e448d0ea9`), reading the same code paths on current main.
Runtime: vLLM 0.27.2.dev1 (source install), torch 2.13.0+cu130, 2x NVIDIA L20, TP=2, Python 3.12.
The issue is scheduling logic, not environment-specific.

### 🐛 Describe the bug

The v1 `LogitsProcessor` interface documents `output_tok_ids` as a live view of each request's generated tokens:

> Key assumption: the `output_tok_ids` list (which is an element of each tuple in `added`) is a reference to the request's running output tokens list; via this reference, the logits processors always see the latest list of generated output tokens.

https://github.com/vllm-project/vllm/blob/6e448d0ea9/vllm/v1/sample/logits_processor/interface.py#L45-L48

Under async scheduling (default-on since 0.27), that assumption only holds conditionally. The worker-side lists are extended with `-1` placeholders each step (real ids stay on the GPU in `prev_sampled_token_ids`):

https://github.com/vllm-project/vllm/blob/6e448d0ea9/vllm/v1/worker/gpu_model_runner.py#L3823-L3846

A repair mechanism exists precisely for this: `set_async_sampled_token_ids` / `update_async_output_token_ids` patch the `-1`s with real ids right before the logits processors run:

https://github.com/vllm-project/vllm/blob/6e448d0ea9/vllm/v1/worker/gpu_input_batch.py#L1030-L1056

But the repair is gated by `needs_output_token_ids` (penalties / bad_words / thinking budget / `logitsprocs_need_output_token_ids`):

https://github.com/vllm-project/vllm/blob/6e448d0ea9/vllm/v1/worker/gpu_input_batch.py#L908-L922

and the logitsprocs contribution to that gate only counts **CLI-passed custom processors**:

https://github.com/vllm-project/vllm/blob/6e448d0ea9/vllm/v1/worker/gpu_model_runner.py#L714-L726

```python
# We currently don't know whether a particular custom logits processor
# uses output token ids so we set this conservatively. ...
logitsprocs_need_output_token_ids=bool(custom_logitsprocs),
```

So there are three under-/over-approximations of the same unknown:

1. **Entry-point plugins** (`vllm.logits_processors` group) are not counted — being fixed by #38199.
2. **Builtin-registered processors are never counted.** The three in-tree builtins happen to only read the list's *length* (`MinTokensLogitsProcessor`), never its *values*, so nothing in-tree notices. Any stateful processor registered via `BUILTIN_LOGITS_PROCESSORS` that reads values silently computes on placeholder garbage: full per-step overhead is paid, results are wrong, and there is no warning or error.
3. Conversely, CLI custom processors that *don't* read values still force the repair path (the conservative `True`), paying an unnecessary sync in async mode.

### Reproduction / real-world impact

We integrated Google DeepMind's SynthID-Text watermarking (via the HF `transformers` processor) as a builtin stateful logits processor on top of `releases/v0.27.1` — the first "reads values" consumer of this contract we know of:

- with `--no-async-scheduling`: 8/8 watermarked outputs detected, mean-g z-score avg **7.86** (range 5.51–11.47), 0 false positives on controls;
- with default async scheduling and identical code: **0/8 detected** (z avg 0.39 ≈ control 0.33), while still paying ~11% per-request latency for the watermark compute — the context window degrades to `[-1, -1, -1, -1]`.

Branch with the processor and a fail-safe guard (refuses to watermark and warns when async scheduling is detected): https://github.com/danbo610/vllm/tree/feature/synthid-watermark (see commit `f18c17a3bb` for the guard).

Minimal repro sketch: register any processor that prints `output_tok_ids` from `apply()` in `BUILTIN_LOGITS_PROCESSORS`, run with defaults (no penalties, no CLI `--logits-processors`), and observe `[-1, -1, ...]`; pass `--no-async-scheduling` (or any CLI custom processor, which flips the flag) and observe real ids.

### Suggested fix

A declarative capability on the interface would solve all three cases at once, mirroring `is_argmax_invariant()`:

```python
class LogitsProcessor(ABC):
    @classmethod
    def needs_output_token_ids(cls) -> bool:
        return True  # conservative default for out-of-tree processors
```

- the three in-tree builtins override it to `False` (no behavior change, removes today's over-approximation for value-agnostic CLI procs);
- `build_logitsprocs` ORs the declarations across builtin + entry-point + CLI processors and feeds `logitsprocs_need_output_token_ids`;
- the `interface.py` "Key assumption" docstring gains a note that the live view is only guaranteed when this capability is declared (or async scheduling is off).

Short of that, amending the docstring plus a startup warning when async scheduling is enabled with non-in-tree processors would at least make the failure mode discoverable — today it is completely silent.

Happy to help validate a fix against the watermarking use case; we can also upstream the SynthID processor itself as a separate PR if there is interest.

### PR DESCRIPTION
**Pull request by kakiuwang-ui**
## Purpose

Fixes #52461.

Under async scheduling (default-on since 0.27) the worker pads each request's
output token id list with `-1` placeholders every step; the real ids are only
patched back in when `logitsprocs_need_output_token_ids` is set. Today that flag
is `bool(custom_logitsprocs)` — i.e. it keys off whether a *CLI* custom
processor is present — which is both an under- and an over-approximation of the
real question ("does any loaded processor read output token *values*?"):

- **Stateful processors registered via `BUILTIN_LOGITS_PROCESSORS` are missed.**
  They silently read `-1` placeholders under async scheduling: full per-step
  overhead is paid, results are wrong, and there is no warning. (The three
  in-tree builtins happen to read only the list *length*, so nothing in-tree
  trips over it today — see the reporter's SynthID-Text watermarking case in the
  issue for a real "reads values" consumer: 8/8 detections with
  `--no-async-scheduling` → 0/8 with async scheduling.)
- **Value-agnostic CLI processors are over-counted**, forcing an unnecessary
  device sync in async mode.

## Change

Add a declarative capability to the interface, mirroring `is_argmax_invariant()`:

```python
@classmethod
def needs_output_token_ids(cls) -> bool:
    return True  # conservative default; out-of-tree processors are correct by default
```

- The three in-tree builtins (`MinP`, `LogitBias`, `MinTokens`) override to
  `False` — they read at most `len(output_tok_ids)`, which is unaffected by
  placeholder *values*.
- `LogitsProcessors` ORs the declaration across every loaded processor (builtin
  + entry-point plugin + CLI) and exposes a single `needs_output_token_ids`.
- `GPUModelRunner` feeds that into `InputBatch` instead of
  `bool(custom_logitsprocs)`. Thinking-budget tracking stays dynamic (unchanged).

Net effect: the async-scheduling repair now runs iff some loaded processor
actually reads token values — fixing the silent-wrong-output case and dropping
the needless sync for value-agnostic processors.

## Not a duplicate of #38199

#38199 targets only the entry-point-plugin gap, via a coarser
`LogitsProcessors.has_custom` flag (true for any non-builtin processor). That
leaves the `BUILTIN_LOGITS_PROCESSORS`-registered stateful case in this issue
still broken, and does not help the value-agnostic over-count. The per-processor
declaration here is a superset covering all three cases; happy to coordinate
landing order with that PR.

## Test

```
python -m pytest tests/v1/logits_processors/test_correctness.py -v
```

- Full file: **42 passed** (includes the CUDA `test_logitsprocs[...]` correctness
  cases; run on an RTX A6000).
- Two added unit tests:
  - `test_needs_output_token_ids_declarations` — base default is `True`, the
    three builtins declare `False`.
  - `test_build_logitsprocs_aggregates_output_token_ids_gate` — the aggregated
    gate is `False` for builtins-only and for a value-agnostic custom processor,
    and `True` once a value-reading custom processor is loaded.

No in-tree model output changes (the builtins keep identical behavior; only the
async-scheduling repair *gating* changes), so model evals are N/A; the
end-to-end correctness win is the reporter's SynthID watermarking measurements
in the issue, which I'd be glad to help re-validate.

## Note

This change was developed with AI assistance. I have reviewed every changed
line, understand the change end-to-end, and ran the tests above myself.

### CODE REVIEW

**Comment by kakiuwang-ui** [comment]
### End-to-end validation under async scheduling

Ran a discriminating A/B on GPU (RTX A6000, `facebook/opt-125m`, `enforce_eager`,
async scheduling on). A probe `LogitsProcessor` registered via
`BUILTIN_LOGITS_PROCESSORS` snapshots each request's `output_tok_ids` from
`apply()` (single-process, `VLLM_ENABLE_V1_MULTIPROCESSING=0`). Only the probe's
`needs_output_token_ids()` declaration differs between runs:

| `needs_output_token_ids()` | observed `output_tok_ids` in `apply()` | outcome |
| --- | --- | --- |
| `True` (this PR's default for out-of-tree procs) | `[5] → [5, 812] → [5, 812, 9] → …` (real ids) | correct |
| `False` (today's behavior for builtin-registered procs) | `[-1] → [-1, -1] → [-1, -1, -1] → …` | silently reads placeholders |

Both runs produce identical output text (the probe only observes), so this
isolates the output-token-id view. The `False` case reproduces the exact silent
failure from the issue; with the declaration wired through the gate, the async
`-1` repair runs and the processor sees real ids.

### Why this is a V1-runner fix (and V2 does not need the same change)

The V2 model runner (`vllm/v1/worker/gpu/model_runner.py`) does not run
custom/plugin logits processors at all: `_get_v2_model_runner_unsupported_features`
lists `"custom logits processors"` as unsupported whenever
`model_config.logits_processors` or a `vllm.logits_processors` entry point is
present (`vllm/config/vllm.py:2447-2450`), so those configs fall back to the V1
runner. The gate this PR fixes lives on the V1 path, which is exactly where such
processors execute. (The e2e forces `VLLM_USE_V2_MODEL_RUNNER=0` only because
`opt-125m` happens to default to V2; a real custom/plugin processor forces V1 on
its own.)

### Unit + regression

`python -m pytest tests/v1/logits_processors/test_correctness.py -v` → **42 passed**
(includes the CUDA `test_logitsprocs[...]` cases), plus the two added unit tests.
`ruff check` / `ruff format --check` clean on the touched files.

**Comment by kakiuwang-ui** [comment]
Good catch — fixed in 48c70dc. I moved `build_logitsprocs` back inside the `with self.jit_warmup_registry.activate():` block (right before the `InputBatch(...)` construction), so any warmup providers a processor's `__init__` registers are captured exactly as they were before this PR. The gate still reads `logitsprocs.needs_output_token_ids` — the only change vs. main is that the built object is now a local instead of an inline argument, and it's constructed within the same context.

No in-tree processor registers warmup providers today, but keeping construction inside the context makes it robust regardless, and avoids reasoning about registry internals.

Thanks for the production numbers (8/8, mean-g z 8.77) — that's a great data point. If a maintainer wants an extra signal before merge, an A/B of your SynthID setup against this branch would be very welcome.

**Comment by kakiuwang-ui** [comment]
Re-validated after the warmup-context fix (48c70dc), on the same RTX A6000 setup:

- `pytest tests/v1/logits_processors/test_correctness.py` → **42 passed** (incl. the CUDA cases) + the two added unit tests.
- V1 async-scheduling e2e A/B still discriminating end-to-end (the probe now goes through `GPUModelRunner.__init__` → `with self.jit_warmup_registry.activate(): build_logitsprocs(...); InputBatch(...)`): `needs_output_token_ids()=True` → real ids `[5] → [5,812] → …`; `False` → `[-1] → [-1,-1] → …`.

So moving `build_logitsprocs` back inside the warmup context is behavior-preserving and the gate still works as intended.

**Comment by kakiuwang-ui** [comment]
cc @njhill — this touches `/vllm/v1/sample` and `/vllm/v1/worker`, which you own in CODEOWNERS. Would appreciate a look when you have a moment.

Summary: under async scheduling, stateful logits processors that read output-token-id *values* saw `-1` placeholders because the `-1` repair gate was derived from `bool(custom_logitsprocs)` (missing builtin/plugin-registered processors, over-triggering for value-agnostic CLI ones). This replaces that heuristic with a declarative `LogitsProcessor.needs_output_token_ids()` (default `True`; the three in-tree builtins override to `False`), OR-aggregated across all loaded processors.

- Reporter @danbo610 reviewed and confirmed the approach, with production numbers under async + SynthID watermarking (8/8 detected, zero FP).
- `tests/v1/logits_processors/test_correctness.py` → 42 passed (incl. CUDA) + 2 new unit tests; async-scheduling e2e A/B posted above.

If it looks reasonable, could you add the `ready` label so full CI can run? Happy to make any changes you'd prefer.

**Comment by kakiuwang-ui** [comment]
@claude review

**Comment by kakiuwang-ui** [comment]
Quick status for reviewers: this has been rebased onto main and remains MERGEABLE (format + DCO green). It's `mrv1-only`, reporter-approved by @danbo610 (with production numbers under async scheduling + SynthID watermarking), and @danbo610 plans to upstream the SynthID processor as a follow-up PR once this enabling fix lands. Whenever a maintainer has a moment — could it get the `ready` label so full CI can run? Happy to make any changes.

**Comment by kakiuwang-ui** [comment]
Rebased onto current `main` (merge `f2cb805c`), conflict resolved.

The conflict was the `InputBatch` construction in `gpu_model_runner.py`. Worth flagging for reviewers: `main` has since **removed the `with self.jit_warmup_registry.activate():` wrapper around that site**, which is what an earlier round of review on this PR asked me to keep `build_logitsprocs` inside. With the wrapper gone the resolution is smaller than before — `build_logitsprocs(...)` is simply hoisted to a local so its aggregate can be read twice:

```python
logitsprocs = build_logitsprocs(...)
self.input_batch = InputBatch(
    ...
    logitsprocs=logitsprocs,
    logitsprocs_need_output_token_ids=logitsprocs.needs_output_token_ids,
    ...
)
```

replacing `logitsprocs_need_output_token_ids=bool(custom_logitsprocs)`. The re-init site in `initialize_kv_cache` already forwards `self.input_batch.logitsprocs_need_output_token_ids`, so it needs no change.

Re-validated on the merge commit (RTX A6000): `pytest tests/v1/logits_processors/test_correctness.py` → **45 passed**, including the CUDA-parametrised cases and the two tests added here. `ruff check` / `ruff format --check` (v0.14.0, the pinned revision) clean on all five changed files.

**Comment by kakiuwang-ui** [comment]
@njhill — a small process ask rather than a review request: could this get the `ready` label?

`pre-run-check` is the only red check here, and it gates on the label (or on the author having ≥4 merged PRs), so this PR has never had a full CI run. Everything else is green: DCO, format, and the branch is MERGEABLE on current `main` after the `f2cb805c` rebase. My other PR (#54065) was in exactly this state until it got `ready` yesterday — `pre-run-check` went green immediately and Buildkite is running now — so I suspect this one is just waiting on the same nudge.

Recap of where it stands: the fix is `mrv1-only` (V2 falls back to V1 whenever custom/plugin logits processors are present), reporter-approved by @danbo610 with production data, and re-validated on the current merge — `tests/v1/logits_processors/test_correctness.py` → 45 passed, plus a V1-forced e2e A/B showing a processor declaring `needs_output_token_ids()` sees real ids while one declaring `False` sees `-1` placeholders.

Happy to run `/ci run` myself once the label is on.
